File: intermediario/leetcode/longest_palindromic_substring/LongestPalindromicsubstring.java


public class LongestPalindromicsubstring {

    public static void main(String[] args) {

        Solution s = new Solution();

        String result = s.longestPalindrome("cbbd");

        System.out.println(result);
    }

}


class Solution {
    public String longestPalindrome(String s) {

        int left_window = 0;
        int right_window = s.length() - 1;
        String polindrome = "";

        while(left_window < s.length()){

            while(right_window > left_window){

                int left = left_window;
                int right = right_window;

                while(left < right){

                    if(s.charAt(left) != s.charAt(right)){
                        break;
                    }

                    left++;
                    right--;

                }

                if(left >= right && polindrome.length() < (right_window - left_window + 1)){
                    polindrome = s.substring(left_window, right_window+1);
                }

                right_window--;
            }
            
            left_window++;
            right_window = s.length() - 1;
        }

        return polindrome;
    }
}

File: intermediario/leetcode/longest_palindromic_substring/longest_palindromic_substring.py
def longestPalindrome(s):
    
    center = 0

    start = 0
    end = 0

    n= len(s)

    while center < len(s):

        if (2 * min(center, n - center - 1) + 1) + 1 < start - end + 1:
            break

        left = center
        right = center + 1

        while left >= 0 and right < len(s) and s[left] == s[right]:
            left-=1
            right+=1

        if (end - start + 1) < (right - 1 - left + 1 - 1):
            start = left + 1
            end = right - 1

        print(start, end, s[start:end+1])


        left = center
        right = center

        while left >= 0 and right < len(s) and s[left] == s[right]:
            left-=1
            right+=1

        if (end - start + 1) < (right - 1 - left + 1 - 1):
            start = left + 1
            end = right - 1
        
        center+=1

        print(start, end, s[start:end+1])

    return s[start:end+1]

s = "babad"

print(longestPalindrome(s))
        



